### PR TITLE
MatrixFree: do not update ghost values of a ghosted src vector

### PR DESCRIPTION
**Pull request by peterrum**
similar to #15865

### CODE REVIEW

**Comment by peterrum** [comment]
For some reasons, the following tests trigger asserts:
```
	594 - matrix_free/ecl_05.mpirun=1.debug (Failed)
	596 - matrix_free/ecl_05.mpirun=1.release (Failed)
	1366 - matrix_free/step-48.mpirun=10.debug (Failed)
	1367 - matrix_free/step-48.mpirun=10.release (Failed)
	1368 - matrix_free/step-48.mpirun=4.debug (Failed)
	1369 - matrix_free/step-48.mpirun=4.release (Failed)
	1373 - matrix_free/step-48c.mpirun=4.debug (Failed)
	1375 - matrix_free/step-48c.mpirun=4.release (Failed)
```

@kronbichler Do you see anything obviously wrong?

**Comment by kronbichler** [comment]
For `step-48`, I would imagine that the error is caused by the fact that we use an `std::vector<LA::dist::Vector>` with the old solution from two previous time steps: https://github.com/dealii/dealii/blob/b7729c72e9520f7193c3e18cead36b6866dba3d5/tests/matrix_free/step-48.cc#L122-L126
In `output_results`, we call https://github.com/dealii/dealii/blob/b7729c72e9520f7193c3e18cead36b6866dba3d5/tests/matrix_free/step-48.cc#L322-L323 but we do not zero out the ghost values, so we might end up having one entry in the `std::vector` having ghost values set and the other not, leading to contradictions down the line. I think this was missed when we cleaned up the behavior of ghosted/non-ghosted vectors in 2016-2018, but the solution should be to call `solution.zero_out_ghost_values()` at the end of `output_results` (both in `tests/matrix_free/step-48.cc`, `tests/matrix_free/step-48b.cc`, `tests_matrix_free/step-48c.cc` and `examples/step-48/step-48.cc`).

**Comment by kronbichler** [comment]
For the `ecl_05` test, I had a quick look at the code and I can't really tell what goes wrong there, it uses some usual code with a single operator evaluation in some strategies, not sure what can go wrong there and why vectors would be affected by the ghost state in the first place. What is the error that you get? Assertion or just wrong result when the vectors are printed?

File: tests/matrix_free/step-48.cc [hunk 1]
```diff
@@ -330,6 +330,7 @@ namespace Step48
                                       norm_per_cell,
                                       QGauss<dim>(fe_degree + 1),
                                       VectorTools::L2_norm);
+    solution.zero_out_ghost_values();
```
Comment (peterrum) [inline tests/matrix_free/step-48.cc:333]: @kronbichler I am zeroing now the ghost values here.

**Comment by peterrum** [comment]
Regarding `matrix_free/ecl_05.mpirun=1.debug`: this tests seems to never have worked on AVX512. This is the ECL test using `FE_Q` written by @bergbauer for CutFEM. It seems that while reading from the global vector the DG path is taken and not the CG path. Generally, I am not sure how much sense it makes to use `FE_Q` with the current interface: compression on the destination vector is disabled.

File: tests/matrix_free/step-48.cc (hunk 1 above)
Comment (bangerth) [reply tests/matrix_free/step-48.cc:333]: Would you have to do the same also in the step-48 tutorial program?

Comment (bangerth) [reply tests/matrix_free/step-48.cc:333]: Separately, though, it isn't clear to me why this is the right place for it. You don't use `solution` any more in the remainder of the function. From a readability perspective, I'd move the call to `zero_out_ghost_values()` to whatever the next place in control flow is that you need the ghost values to be zero.

**Comment by bangerth** [comment]
@peterrum @kronbichler Is this still on your radar?

**Comment by peterrum** [comment]
> Is this still on your radar?

Should be ready.

**Comment by kronbichler** [comment]
```

matrix_free/stokes_computation.mpirun=4.debug: RUN failed. ------ Additional output on stdout/stderr:


--------------------------------------------------------
An error occurred in line <1647> of file </jenkins/workspace/dealii-mpi_PR-15959/include/deal.II/lac/la_parallel_vector.h> in function
    Number dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::local_element(dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::size_type) const [with Number = double; MemorySpace = dealii::MemorySpace::Host; dealii::LinearAlgebra::distributed::Vector<Number, MemorySpace>::size_type = unsigned int]
The violated condition was: 
    local_index < locally_owned_size() || vector_is_ghosted == true
Additional information: 
    You tried to read a ghost element of this vector, but it has not
    imported its ghost values.
```
There are two variants of this Stokes computation, the second one in the mg global coarsening part. Can you please have a look?

**Comment by peterrum** [comment]
> There are two variants of this Stokes computation, the second one in the mg global coarsening part. Can you please have a look?

@kronbichler  I have updated the tests!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**